Documentation/design-docs/standalone-gc-eventing.md

```diff
@@ -0,0 +1,212 @@
+responsible for routing the event to any appropriate subsystems (ETW, LTTNG, EventPipe) and the GC has
+no knowledge of what it is going to do with events that we give it.
+
+Given that we have validated that an event is enabled (through `GCEventStatus::IsEventEnabled`), we can
+fire the event using this new method on `IGCToCLR`:
+
+```c++
+void IGCToCLR::FireKnownEvent(
+    /* IN */ void* eventHandle,
+    /* IN */ void** eventData,
+    /* IN */ uint32_t eventDataCount,
+    /* IN */ const GUID* pActivityId = nullptr,
+    /* IN */ const GUID* pRelatedActivityId = nullptr
+);
+```
+
+The four arguments of this function are:
+1. `eventHandle` - An opaque handle to the EE's representation of the event being fired.
+2. `eventData` - An array of pointers that contains the payload of the event being fired.
+Each `GCEventData` is a tuple of a data pointer and size, so it can be used to describe arbitrary data.
+3. `eventDataCount` - The size of the `eventData` array.
+4. `pActivityId` and `pRelatedActivityId` - If provided, passes the activity ID and related activity ID
+verbatim to the underlying event implementation.
+
+The EE will use this information to determine 1) if this event truly is enabled and, 2) if it is,
+forwarding it to the apporopriate underlying event implementation, be it EventPipe, ETW, or LTTNG.
+The passed-in eventHandle will be a handle to an object from which the EE can get information about
+the version of the event and other useful metadata.
+
+## Getting Informed of Changes to Event State
+
+There are three mechanisms by which CoreCLR is able to log events: EventPipe, ETW, and LTTNG. When it
+comes to changing the state of events, EventPipe and ETW both allow users to attach callbacks that are
+invoked whenever events are enabled or disabled. For these two mechanisms, it is sufficient to use
+this callback mechanism to call `IGCHeap::{Enable/Disable}Events` from within such a callback in order
+to inform the GC of changes to tracing state.
+
+LTTNG does not have such a mechanism. In order to observe changes in the eventing state, LTTNG must be
+polled periodically. For this purpose, when tracing using LTTNG, a new thread will be spawned by the EE
+whose responsibility is to periodically poll LTTNG and invoke the GC whenever it detects changes in the
+eventing state.
+
+## Adding New Events
+
+It is useful for a standalone GC to be able to fire events that the EE was not previously aware of.
+For example, it is useful for a GC developer to add some event-based instrumentation to the GC, especially
+when testing new features and ensuring that they work as expected.
+
+While it is possible for some eventing implementations to receive events that are created at runtime, not all
+eventing implementations (particularly LTTNG) are not flexible enough for this. In order to accomodate new
+events, another method is added to `IGCToCLR`:
+
+```c++
+void IGCToCLR::FireCustomEvent(
```

Comment:
It seems that majority of events will turn into custom events over time with this design.

Would it make more sense to have just one FireEvent method that is used for all events, and instead allow GC to define handles for "custom events" ? This is related to my other question how the GC gets handles for the well-known events.

Reply:
I've been pondering this question for a week and I think that the goal of this particular API is that it is an escape hatch for standalone GCs that want to add their own events for the purpose of validating some property of that standalone GC. If we (as the CLR GC developers) want to add a new event that we intend to be useful when *not* running standalone, I think that we should bump the minor version number of the EE/GC interface and add it as a well-known event.

I don't think it's desirable to ever be in a situation where the default GC (i.e. the one that you get when you build this repo and the one that we ship) fires "custom" events. Custom events pass the buck entirely to PerfView. I'm imagining that an author of a custom GC would use this API in order to fire events that the EE will *never* be aware of - the details of my custom GC that I do not intend to ever bring to the default GC.

Reply:
> I think that the goal of this particular API is that it is an escape hatch for standalone GCs

Make sense.

> I don't think it's desirable to ever be in a situation where the default GC

It may be worth mentioning this in the design document.

Reply:
will do!

Reply:
I think it's worth illustrating the usage scenarios here - if one adds a custom event, what do you expect to happen? I would expect clr.dll to certainly NOT reject loading this clrgc.dll. if that's the case what does "bump the minor version number" entail? tools would interpret what they can, eg, PerfView will just specify the event layout like it does today and obviously we'd update the layout as we add more events so whether an event is viewed as "known" or "custom" makes no difference. is there another tooling scenario that would differentiate between "known" or "custom" events?

Reply:
I'll add it to the document, but firing a known event that the EE does not known about will cause the event to be dropped. PerfView will never see it because it will never have been emitted in the first place.

If you add a custom event, it'll work with any version of the EE - `FireCustomEvent` takes a byte array as the event payload, so it can be used to fire any kind of event. The tooling scenario would be that, if you adding a custom event, you'll need to instruct PerfView how to parse the binary event data that you gave to `FireCustomEvent`.

Reply:
PerfView is always instructed of the new event layout regardless of how this event was added. so my question still stands - what exactly is the benefit of differentiating between a known and a custom event? you mentioned bumping up the minor version. what does that effect? it doesn't affect PerfView 'cause whenever we add a new event we just add its layout to PerfView anyway.

Reply:
I think what @swgillespie is getting at is that you need to bump the version if you add a new event to the GC that is not a "custom" event because changes to the EE are required to support non-custom events.  New events that are not custom events are those that are described in the manifest and follow the standard conventions followed by ETW, LTTng and EventPipe.  To support new non-custom events, changes to the EE are required.  Custom events do not require any changes to the EE, but come with downsides - they are not really interpretable by any tool other than PerfView whereas non-custom events can be consumed by native tools such as WPA, LTTng/BabelTrace, etc.

Custom events are really about making sure that standalone GCs can emit instrumentation without a huge burden of having to directly consume platform loggers.  At the same time, custom events really belong to the developer of the standalone GC and not to the .NET Core stack so it would be philosophically OK to require standalone GC implementers to directly log to platform loggers, but from a pragmatic POV that's painful.